File: app/src/main/java/com/musicplayer/SocyMusic/ui/all_songs/SongListAdapter.java
package com.musicplayer.SocyMusic.ui.all_songs;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.musicplayer.SocyMusic.data.Song;
import com.musicplayer.musicplayer.R;
import java.util.List;

/**
 * Custom adapter for SongsData related actions
 */
public class SongListAdapter extends RecyclerView.Adapter<SongHolder> implements Filterable {
    private final Context context;
    private List<Song> allSongs;
    private ItemClickListener clickListener;
    private List<Song> filteredList;

    public SongListAdapter(Context context, List<Song> allSongs) {
        super();
        this.allSongs = allSongs;
        this.context = context;
        filteredList = allSongs;

    }

    @NonNull
    @Override
    public SongHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View itemView = inflater.inflate(R.layout.list_item_all_songs, parent, false);
        return new SongHolder(itemView, clickListener);
    }

    @Override
    public void onBindViewHolder(@NonNull SongHolder holder, int position) {
        Song song = filteredList.get(position);
        holder.bind(song);
    }

    @Override
    public int getItemCount() {
        return filteredList == null ? 0 : filteredList.size();
    }

    public void setOnItemClickListener(ItemClickListener clickListener) {
        this.clickListener = clickListener;
    }

    public void setAllSongs(List<Song> allSongs) {
        this.allSongs = allSongs;
    }

    // This filter will select the searched song and display it
    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence charSequence) {
                String charString = charSequence.toString();

                if (charString.isEmpty()) {
                    filteredList = allSongs;
                } else {
                    filteredList.clear();
                    for (Song data : allSongs) {
                        //In this loop, you'll filter through originalData and compare each item to charSequence.
                        //If you find a match, add it to your new ArrayList
                        if (data.getTitle().toLowerCase().contains(charSequence.toString().toLowerCase())) {
                            filteredList.add(data);
                        }
                    }
                }

                FilterResults results = new FilterResults();
                results.values = filteredList;
                results.count = filteredList.size();
                return results;
            }

            @Override
            protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
                filteredList = (List<Song>) filterResults.values;
                notifyDataSetChanged();
            }
        };
    }

    public interface ItemClickListener {
        void onItemClick(int position, View view);

        boolean onItemLongClick(int position, View view);
    }

}

File: app/src/main/kotlin/com/musicplayer/openmusic/ui/all_songs/SongListAdapter.kt
package com.musicplayer.openmusic.ui.all_songs

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import androidx.recyclerview.widget.RecyclerView
import com.musicplayer.openmusic.data.Song
import com.musicplayer.musicplayer.R
import java.util.*

/**
 * Custom adapter for SongsData related actions
 */
class SongListAdapter(private val context: Context, private var allSongs: MutableList<Song>) :
    RecyclerView.Adapter<SongHolder>(), Filterable {
    private var clickListener: ItemClickListener? = null
    private var filteredList: MutableList<Song>?

    init {
        filteredList = allSongs
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SongHolder {
        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val itemView = inflater.inflate(R.layout.list_item_all_songs, parent, false)
        return SongHolder(itemView, clickListener!!)
    }

    override fun onBindViewHolder(holder: SongHolder, position: Int) {
        val song = filteredList!![position]
        holder.bind(song)
    }

    override fun getItemCount(): Int {
        return if (filteredList == null) 0 else filteredList!!.size
    }

    fun setOnItemClickListener(clickListener: ItemClickListener) {
        this.clickListener = clickListener
    }

    fun setAllSongs(allSongs: MutableList<Song>) {
        this.allSongs = allSongs
    }

    // This filter will select the searched song and display it
    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(charSequence: CharSequence): FilterResults {
                val charString = charSequence.toString()
                if (charString.isEmpty()) {
                    filteredList = allSongs
                } else {
                    for (data in allSongs) {
                        if (data.title.lowercase(Locale.getDefault()).contains(
                                charSequence.toString().lowercase(
                                    Locale.getDefault()
                                )
                            )
                        ) {
                            filteredList!!.add(data)
                        }
                    }
                }
                val results = FilterResults()
                results.values = filteredList
                results.count = filteredList!!.size
                return results
            }

            override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {
                filteredList = filterResults.values as MutableList<Song>
                notifyDataSetChanged()
            }
        }
    }

    interface ItemClickListener {
        fun onItemClick(position: Int, view: View)
        fun onItemLongClick(position: Int, view: View): Boolean
    }
}